README
Pod-Loom version {{$version}}, released {{$date}}


Pod::Loom extracts all the POD sections from Perl code, passes the POD
to a template that may reformat it in various ways, and then returns a
copy of the code with the reformatted POD at the end.



INSTALLATION

To install this module, run the following commands:

{{ $t->build_instructions('    ') }}



DEPENDENCIES

{{ $t->dependency_list }}



CHANGES
    Here's what's new in version {{$version}} of Pod-Loom:
    (See the file "Changes" for the full revision history.)

{{$changes}}



COPYRIGHT AND LICENCE

{{ local $_ = $zilla->license->notice; chomp; $_ }}


Changes
Revision history for Pod-Loom

0.08   NOT
	- Change 10-weave.t to work around bug in MetaCPAN indexer

0.07   2014-03-22
	- Make POD encoded in Latin-1 work with newer versions of
	  Pod::Eventual, which treat strings as UTF-8 by default

0.06   2013-01-12
	- Don't insert __END__ if a file contains only POD

0.05   2011-10-15
	- Handle =encoding directive
	- Added bugtracker attribute to Template::Default

0.04   2010-12-11
	- Use S<C<< >>> instead of \xA0
	  (It's S<< C<< >> >> that Pod::Simple misinterprets.)

0.03   2010-03-06
	- If the template returns no POD, don't add a =cut
	- Include filename in error messages and warnings
	  Added the error & warning methods to Template
	- Added SUBROUTINES section to Template::Default
	- Use \xA0 instead of S<> with C<>
	  (search.cpan.org's pod2html doesn't seem to like S<C<>>)

0.02   2009-10-20
	- INCOMPATIBLE CHANGE
	  Templates should now provide a _build_sections method
	  instead of a default value for sections
	- Added the Extender role for custom templates
	- Added support for Pod::Loom-group_COMMAND to Template

0.01   2009-10-11
	- Initial release


lib/Pod/Loom.pm
#---------------------------------------------------------------------
package Pod::Loom;
#
# Copyright 2009 Christopher J. Madsen
#
# Author: Christopher J. Madsen <perl@cjmweb.net>
# Created: October 6, 2009
#
# This program is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either the
# GNU General Public License or the Artistic License for more details.
#
# ABSTRACT: Weave pseudo-POD into real POD
#---------------------------------------------------------------------

use 5.008;
our $VERSION = '0.08';
# This file is part of {{$dist}} {{$dist_version}} ({{$date}})

use Moose 0.65; # attr fulfills requires
use Carp qw(croak);
use PPI ();
use String::RewritePrefix ();

#=====================================================================
{
  package Pod::Loom::_EventCounter;
  our @ISA = 'Pod::Eventual';
  sub new {
    require Pod::Eventual;
    my $events = 0;
    bless \$events => shift;
  }

  sub handle_event { ++${$_[0]} }
  sub events { ${ +shift } }
}

#=====================================================================
# Package Pod::Loom:

has template => (
  is      => 'rw',
  isa     => 'Str',
  default => 'Default',
);
#=====================================================================

=method weave

    $new_doc = $loom->weave(\$doc, $filename, $data);

This method does all the work (see L</"DESCRIPTION">).  You pass it a
reference to a string containing Perl code mixed with POD.  (This
string is not modified.)  It returns a new string containing the
reformatted POD moved to the end of the code.  C<$doc> should contain
raw bytes (i.e. UTF8 flag off).  If C<$doc> is encoded in something
other than Latin-1, it must contain an C<=encoding> directive
specifying the encoding.  C<$new_doc> will likewise contain raw bytes
in the same encoding as C<$doc>.

The C<$filename> is used for error messages.  It does not need to
actually exist on disk.

C<$data> is passed as the only argument to the template class's
constructor (which must be named C<new>).  Pod::Loom does not inspect
it, but for consistency and compatibility between templates it should
be a hashref.

=cut

sub weave
{
  my ($self, $docRef, $filename, $data) = @_;

  my $ppi = PPI::Document->new($docRef);

  my $sourcePod = join("\n", @{ $ppi->find('PPI::Token::Pod') || [] });

  $ppi->prune('PPI::Token::Pod');

  croak "Can't use Pod::Loom on $filename: there is POD inside string literals"
      if $self->_has_pod_events("$ppi");

=diag C<< Can't use Pod::Loom on %s: there is POD inside string literals >>

You have POD commands inside a string literal (probably a here doc).
Since Pod::Loom moves all POD to the end of the file, running it on
your program would change its behavior.  Move the POD outside the
string, or quote any equals sign at the beginning of a line so it no
longer looks like POD.

=cut

  # Determine the template to use:
  my $templateClass = $self->template;

  if ($sourcePod =~ /^=for \s+ Pod::Loom-template \s+ (\S+)/mx) {
    $templateClass = $1;
  }

  $templateClass = String::RewritePrefix->rewrite(
    {'=' => q{},  q{} => 'Pod::Loom::Template::'},
    $templateClass
  );

  # Instantiate the template and let it weave the new POD:
  croak "Invalid class name $templateClass"
      unless $templateClass =~ /^[:_A-Z0-9]+$/i;
  eval "require $templateClass;" or croak "Unable to load $templateClass: $@";

=diag C<< Invalid class name %s >>

A template name may only contain ASCII alphanumerics and underscore.

=diag C<< Unable to load %s: %s >>

Pod::Loom got an error when it tried to C<require> your template class.

=cut

  my $template = $templateClass->new($data);

  my $newPod = $template->weave(\$sourcePod, $filename);
  $newPod =~ s/(?:\s*\n=cut)*\s*\z/\n\n=cut\n/; # ensure it ends with =cut
  $newPod = '' if $newPod =~ /^\s*=cut$/;       # if it's blank, ignore it

  # Plug the new POD back into the code:

  my $end = do {
    my $end_elem = $ppi->find('PPI::Statement::Data');

    unless ($end_elem) {
      $end_elem = $ppi->find('PPI::Statement::End');

      # If there's nothing after __END__, we can put the POD there:
      if (not $end_elem or (@$end_elem == 1 and
                            $end_elem->[0] =~ /^__END__\s*\z/)) {
        $end_elem = [];
      } # end if no significant text after __END__
    } # end unless found __DATA__

    @$end_elem ? join q{}, @$end_elem : undef;
  };

  $ppi->prune('PPI::Statement::End');
  $ppi->prune('PPI::Statement::Data');

  my $docstr = $ppi->serialize;
  $docstr =~ s/\n*\z/\n/;       # ensure it ends with one LF

  return $newPod if $docstr eq "\n" and not defined $end; # Pure POD file

  return defined $end
      ? "$docstr\n$newPod\n$end"
      : "$docstr\n__END__\n\n$newPod";
} # end weave_document

#---------------------------------------------------------------------
sub _has_pod_events
{
  my $pe = Pod::Loom::_EventCounter->new;
  # We can't use read_string, because that treats the string as
  # encoded in UTF-8, for which some byte sequences aren't valid.
  open my $handle, '<:encoding(iso-8859-1)', \$_[1]
      or die "error opening string for reading: $!";
  $pe->read_handle($handle);

  $pe->events;
} # end _has_pod_events

#=====================================================================
# Package Return Value:

no Moose;
__PACKAGE__->meta->make_immutable;
1;

__END__

=for Pod::Loom-insert_before
WARNING
SYNOPSIS

=head1 WARNING

This code is still in flux.  Use it at your own risk, and be prepared
to adapt to changes.  The POD syntax should be fairly stable, but if
you write your own templates, they might need to change.

=head1 SYNOPSIS

  use Pod::Loom;

  my $document = ...; # Text of Perl program including POD
  my $filename = "filename/of/document.pm"; # For messages
  my %data = ...; # Configuration required by template

  my $loom = Pod::Loom->new(template => 'Custom');
  my $new_doc = $loom->weave(\$document, $filename, \%data);

=head1 DESCRIPTION

Pod::Loom extracts all the POD sections from Perl code, passes the POD
to a template that may reformat it in various ways, and then returns a
copy of the code with the reformatted POD at the end.

A template may convert non-standard POD commands like C<=method> and
C<=attr> into standard POD, reorder sections, and generally do
whatever it likes to the POD.

The document being reformatted can specify the template to use with a
line like this:

  =for Pod::Loom-template TEMPLATE_NAME

Otherwise, you can specify the template in the Pod::Loom constructor:

  $loom = Pod::Loom->new(template => TEMPLATE_NAME);

TEMPLATE_NAME is automatically prefixed with C<Pod::Loom::Template::>
to form a class name.  If you want to use a template outside that
namespace, prefix the class name with C<=> to indicate that.


=for Pod::Loom-sort_method
new

=method new

  $loom = Pod::Loom->new(template => TEMPLATE_NAME);

Constructs a new Pod::Loom.  The C<template> parameter is optional; it
defaults to C<Default> (meaning L<Pod::Loom::Template::Default>).


=head1 REQUIREMENTS OF A TEMPLATE CLASS

A template class must have a constructor named C<new> and a method
named C<weave> that matches the one in L<Pod::Loom::Template>.  It
should be in the C<Pod::Loom::Template::> namespace (to make it easy
to specify the template name), but it does not need to be a subclass
of Pod::Loom::Template.


=head1 DIAGNOSTICS

Pod::Loom may generate the following error messages, in addition to
whatever errors the template class generates.


=head1 DEPENDENCIES

Pod::Loom depends on L<Moose>, L<Pod::Eventual>, L<PPI>, and
L<String::RewritePrefix>, which can be found on CPAN.  The template
class may have additional dependencies.


lib/Pod/Loom/Parser.pm
#---------------------------------------------------------------------
package Pod::Loom::Parser;
#
# Copyright 2009 Christopher J. Madsen
#
# Author: Christopher J. Madsen <perl@cjmweb.net>
# Created:  6 Oct 2009
#
# This program is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either the
# GNU General Public License or the Artistic License for more details.
#
# ABSTRACT: Subclass Pod::Eventual for Pod::Loom
#---------------------------------------------------------------------

use 5.008;
our $VERSION = '0.05';
# This file is part of {{$dist}} {{$dist_version}} ({{$date}})

use strict;
use warnings;

use Encode qw(find_encoding);
use Pod::Eventual ();
our @ISA = qw(Pod::Eventual);
#---------------------------------------------------------------------

=head1 DEPENDENCIES

Pod::Loom::Parser requires L<Pod::Eventual>, which can be found on CPAN.

=for Pod::Loom-sort_method
new

=method new

  $parser = Pod::Loom::Parser->new(\@collect_commands);

Constructs a new Pod::Loom::Parser.  You pass it an arrayref of the
POD commands at which the POD should be chopped.

=cut

sub new
{
  my ($class, $collectCommands) = @_;

  my %collect = map { $_ => [] } @$collectCommands;
  my %groups  = map { $_ => {} } @$collectCommands;

  bless {
    collect => \%collect,
    dest    => undef,
    groups  => \%groups,
  }, $class;
} # end new

#---------------------------------------------------------------------
sub _handle_encoding
{
  my ($self, $event) = @_;

  my $encoding = $event->{content};

  $encoding =~ s/^\s+//;
  $encoding =~ s/\s+\z//;

  my $e = find_encoding($encoding)
      or die "Invalid =encoding $encoding at line $event->{start_line}\n";

  if (defined $self->{encoding}) {
    return if $e->name eq $self->{encoding}->name;
    die "Conflicting =encoding directive at line $event->{start_line}\n";
  }

  $self->{encoding} = $e;
} # end _handle_encoding

#---------------------------------------------------------------------
sub handle_event
{
  my ($self, $event) = @_;

  my $dest = $self->{dest};

  if ($event->{type} eq 'command') {
    my $cmd = $event->{command};
    return if $cmd eq 'cut';

    return $self->_handle_encoding($event) if $cmd eq 'encoding';

    # See if this changes the output location:
    my $collector = $self->{collect}{ $cmd };

    if (not $collector and $cmd =~ /^(\w+)-(\S+)/ and $self->{collect}{$1}) {
      $collector = $self->{collect}{$cmd} = [];
      $self->{groups}{$1}{$2} = 1;
    } # end if new group

    # Special handling for Pod::Loom sections:
    if ($cmd =~ /^(begin|for)$/ and
        $event->{content} =~ s/^\s*(Pod::Loom\b\S*)\s*//) {
      $collector = ($self->{collect}{$1} ||= []);
      if ($cmd eq 'for') {
        push @$collector, $event->{content};
        return;
      }
      undef $cmd;
    } elsif ($cmd eq 'end' and
             $event->{content} =~ /^\s*Pod::Loom\b/) {
      # Handle =end Pod::Loom:
      $self->{dest} = undef;
      return;
    }

    # Either set output location, or make sure we have one:
    if ($collector) {
      push @$collector, '';
      $dest = $self->{dest} = \$collector->[-1];
    } else {
      die "=$cmd used too soon at line $event->{start_line}\n" unless $dest;
    }

    if ($cmd) {
      $$dest .= "=$cmd";
      $$dest .= ' ' unless $event->{content} =~ /^\n/;
    }
  } # end if command event

  $$dest .= $event->{content};
} # end handle_event

#---------------------------------------------------------------------
sub handle_blank
{
  my ($self, $event) = @_;

  if ($self->{dest}) {
    $event->{type} = 'text';
    $self->handle_event($event);
  }
} # end handle_blank
#---------------------------------------------------------------------

=method collected

  $hashRef = $parser->collected;

This returns the POD chunks that the document was chopped into.  There
is one entry for each of the C<@collect_commands> that were passed to
the constructor.  The value is an arrayref of strings, one for each
time that command appeared in the document.  Each chunk contains all
the text from the command up to (but not including) the command that
started the next chunk.  Chunks appear in document order.

If one of the commands did not appear in the document, its value will
be an empty arrayref.

In addition, any POD targeted to a format matching C</^Pod::Loom\b/>
will be collected under the format name.

=cut

sub collected
{
  my ($self) = @_;

  my $collected = $self->{collect};
  my $encoding  = $self->{encoding} ||= find_encoding('iso-8859-1');

  unless ($self->{collect_decoded}++) {
    for my $array (values %$collected) {
      for my $value (@$array) {
        $value = $encoding->decode($value);
      }
    }
  }

  $collected;
} # end collected

#---------------------------------------------------------------------

=method encoding

  $encoding = $parser->encoding;

This returns the encoding that was used for the document as an
L<Encode> object.  If no encoding was explicitly defined, then the
default Latin-1 encoding is returned.

=cut

sub encoding { shift->{encoding} ||= find_encoding('iso-8859-1') }
#---------------------------------------------------------------------

=method groups

  $hashRef = $parser->groups;

This returns a hashref with one entry for each of the
C<@collect_commands>.  Each value is a hashref whose keys are the
categories used with that command.  For example, if C<attr> was a
collected command, and the document contained these entries:

  =attr-foo attr1
  =attr-bar attr2
  =attr-foo attr3
  =attr attr4

then C<< keys %{ $parser->groups->{attr} } >> would return C<bar> and
C<foo>.  (The C<=attr> without a category does not get an entry in
this hash.)

=cut

sub groups { shift->{groups} }

#=====================================================================
# Package Return Value:

1;

__END__

=head1 SYNOPSIS

  use Pod::Loom::Parser;

  my $parser = Pod::Loom::Parser->new( ['head1'] );
  $parser->read_file('lib/Foo/Bar.pm');
  my $collectedHash = $parser->collected;

  foreach my $block (@{ $collectedHash->{head1} }) {
    printf "---\n%s\n", $block;
  }

=head1 DESCRIPTION

Pod::Loom::Parser is a subclass of L<Pod::Eventual> intended for use
by L<Pod::Loom::Template>.  It breaks the POD into chunks based on a
list of POD commands.  Each chunk begins with one of the commands, and
contains all the POD up until the next selected command.

The commands do not need to be valid POD commands.  You can invent
commands like C<=attr> or C<=method>.

=head1 METHODS

See L<Pod::Eventual> for the C<read_handle>, C<read_file>, and
C<read_string> methods, which you use to feed POD into the parser.


lib/Pod/Loom/Role/Extender.pm
#---------------------------------------------------------------------
package Pod::Loom::Role::Extender;
#
# Copyright 2009 Christopher J. Madsen
#
# Author: Christopher J. Madsen <perl@cjmweb.net>
# Created: October 16, 2009
#
# This program is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either the
# GNU General Public License or the Artistic License for more details.
#
# ABSTRACT: Role to simplify extending a template
#---------------------------------------------------------------------

use 5.008;
our $VERSION = '0.03';
# This file is part of {{$dist}} {{$dist_version}} ({{$date}})

use Moose::Role;
#---------------------------------------------------------------------

=method remap_sections

If your class provides C<remap_sections>, it should return a hashref
keyed by section title.  The values should be arrayrefs of section
titles.  Each section in the hash will be replaced by the listed
sections.  You can use this to insert or remove sections from the
template you're extending.

=cut

around _build_sections => sub {
  my $orig = shift;
  my $self = shift;

  my $sections = $self->$orig(@_);

  my $remap = $self->can('remap_sections');

  if ($remap) {
    $remap = $self->$remap;

    $sections = [ map { $remap->{$_} ? @{$remap->{$_}} : $_ } @$sections ];
  } # end if remap

  $sections;
}; # end around _build_sections
#---------------------------------------------------------------------

=method additional_commands

If your class provides C<additional_commands>, it should return an
arrayref just like C<collect_commands>.  This list will be merged with
the list of C<collect_commands> from the template being extended.

=cut

around collect_commands => sub {
  my $orig = shift;
  my $self = shift;

  my $commands = $self->$orig(@_);

  if (my $additional = $self->can('additional_commands')) {
    my %command = map { $_ => undef } @$commands;

    $command{$_} = undef for @{ $self->$additional };

    $commands = [ keys %command ];
  } # end if additional_commands

  $commands;
}; # end around collect_commands

#=====================================================================
# Package Return Value:

1;

__END__

=head1 SYNOPSIS

  use Moose;
  extends 'Pod::Loom::Template::Default';
  with 'Pod::Loom::Role::Extender';

  sub remap_sections { {
    AUTHOR => [qw[ AUTHOR ACKNOWLEDGMENTS ]],
  } }

  sub section_ACKNOWLEDGMENTS ...

=head1 DESCRIPTION

The Extender role simplifies creating a custom Pod::Loom template.
You should not use this for templates uploaded to CPAN, because you
can only use it once per template.  It's intended for creating a
custom template for a distribution.

=head1 METHODS

Your template class may provide any or all of the following methods.
(If you don't provide any of these methods, then there's no point in
using Extender.)


lib/Pod/Loom/Template/Identity.pm
#---------------------------------------------------------------------
package Pod::Loom::Template::Identity;
#
# Copyright 2009 Christopher J. Madsen
#
# Author: Christopher J. Madsen <perl@cjmweb.net>
# Created:  7 Oct 2009
#
# This program is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either the
# GNU General Public License or the Artistic License for more details.
#
# ABSTRACT: Do-nothing template for Pod::Loom
#---------------------------------------------------------------------

use 5.008;
our $VERSION = '0.03';
# This file is part of {{$dist}} {{$dist_version}} ({{$date}})

use strict;
use warnings;

=for Pod::Loom-sort_method
new

=method new

  $template = Pod::Loom::Template::Identity->new($data);

A template must have a constructor named C<new> that takes one argument.
(In this case, we discard it.)

=cut

sub new
{
  bless {}, shift;
} # end new
#---------------------------------------------------------------------

=method weave

  $new_pod = $template->weave(\$old_pod, $filename);

A template must also have a C<weave> method that returns the new POD.
The Identity template simply returns the POD unchanged.

=cut

sub weave
{
  my ($self, $podRef, $filename) = @_;

  $$podRef;
} # end weave

#=====================================================================
# Package Return Value:

1;

__END__

=head1 DESCRIPTION

This is the simplest possible template for L<Pod::Loom>.  It does
absolutely nothing to the collected POD.  The result is simply to
collect all POD sections and move them to the end of the file.

It demonstrates that a Pod::Loom template does not have to be a
subclass of L<Pod::Loom::Template>, and doesn't even need to use L<Moose>.

=for Pod::Loom-omit
CONFIGURATION AND ENVIRONMENT


lib/Pod/Loom/Template/Default.pm
#---------------------------------------------------------------------
package Pod::Loom::Template::Default;
#
# Copyright 2009 Christopher J. Madsen
#
# Author: Christopher J. Madsen <perl@cjmweb.net>
# Created:  6 Oct 2009
#
# This program is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either the
# GNU General Public License or the Artistic License for more details.
#
# ABSTRACT: Default template for Pod::Loom
#---------------------------------------------------------------------

use 5.008;
our $VERSION = '0.05';
# This file is part of {{$dist}} {{$dist_version}} ({{$date}})

use Moose;
extends 'Pod::Loom::Template';

use Pod::Loom::Template '%E';
#=====================================================================

=attr sort_attr

This should be 0, 1, or an arrayref.  If non 0, attributes will be
sorted.  An arrayref lists the attributes that should come first, with
the remaining attributes in alphabetical order.
See L<Pod::Loom::Template/"Pod::Loom-sort_COMMAND">.

=attr sort_diag

Just like C<sort_attr>, but for diagnostic messages.

=attr sort_method

Just like C<sort_attr>, but for methods.

=attr sort_sub

Just like C<sort_attr>, but for subroutines.

=cut

has qw(sort_attr   is ro), isa => 'Int | ArrayRef[Str]';
has qw(sort_diag   is ro), isa => 'Int | ArrayRef[Str]';
has qw(sort_method is ro), isa => 'Int | ArrayRef[Str]';
has qw(sort_sub    is ro), isa => 'Int | ArrayRef[Str]';

sub collect_commands
{
  [ qw(head1 attr method sub diag) ];
} # end collect_commands

#---------------------------------------------------------------------
# Don't forget to update DESCRIPTION and sort_method if changing this:
our @sections =
  (qw(NAME VERSION SYNOPSIS DESCRIPTION ATTRIBUTES METHODS SUBROUTINES
      * DIAGNOSTICS),
   'CONFIGURATION AND ENVIRONMENT',
   qw(DEPENDENCIES INCOMPATIBILITIES),
   'BUGS AND LIMITATIONS',
   'AUTHOR', 'COPYRIGHT AND LICENSE', 'DISCLAIMER OF WARRANTY');

sub _build_sections { \@sections };
#---------------------------------------------------------------------

=attr abstract

The abstract for the module.  Required by NAME.

=attr module

The name of the module.
Required by NAME and CONFIGURATION AND ENVIRONMENT.

=method section_NAME

  <module> - <abstract>

=cut

has qw(abstract is ro  isa Str);
has qw(module   is ro  isa Str);

sub section_NAME
{
  my ($self, $title) = @_;

  my ($module, $abstract) = $self->required_attr($title, qw(module abstract));

  "=head1 $title\n\n$module - $abstract\n";
} # end section_NAME
#---------------------------------------------------------------------

=attr version

The version number of the module.  Used by VERSION.

=attr version_desc

The complete text of the VERSION section. Used by VERSION.

=method section_VERSION

  <version_desc>

Or, if L</"version_desc"> is not set:

  version <version>

If neither version_desc nor L</"version"> is set, no VERSION section
will be added.

=cut

has qw(version      is ro  isa Str);
has qw(version_desc is ro  isa Str);

sub section_VERSION
{
  my ($self, $title) = @_;

  if ($self->version_desc) {
    return "=head1 $title\n\n$E{$self->version_desc}\n";
  }

  my $version = $self->version;

  return "=head1 $title\n\nversion $version\n" if defined $version;

  '';                           # Otherwise, omit VERSION
} # end section_VERSION
#---------------------------------------------------------------------

=method section_ATTRIBUTES

If the original document contains any C<=attr> commands, they will be
converted to C<=head2> commands in this section.  If there are no
attributes, no ATTRIBUTES section will be added.

=cut

sub section_ATTRIBUTES
{
  my $self = shift;

  $self->joined_section(attr => 'head2', @_);
} # end section_ATTRIBUTES
#---------------------------------------------------------------------

=method section_METHODS

This is just like ATTRIBUTES, except it gathers C<=method> entries.

=cut

sub section_METHODS
{
  my $self = shift;

  $self->joined_section(method => 'head2', @_);
} # end section_METHODS

#---------------------------------------------------------------------

=method section_SUBROUTINES

This is just like ATTRIBUTES, except it gathers C<=sub> entries.

=cut

sub section_SUBROUTINES
{
  my $self = shift;

  $self->joined_section(sub => 'head2', @_);
} # end section_SUBROUTINES

#---------------------------------------------------------------------
sub override_section
{
  my ($self, $title) = @_;

  return ($title eq 'ATTRIBUTES' or
          $title eq 'DIAGNOSTICS' or
          $title eq 'METHODS' or
          $title eq 'SUBROUTINES');
} # end override_section
#---------------------------------------------------------------------

=method section_DIAGNOSTICS

If the original document contains any C<=diag> commands, they will be
converted to an C<=item> list in this section.  If there are no
diagnostics, no DIAGNOSTICS section will be added.

=cut

sub section_DIAGNOSTICS
{
  my $self = shift;

  $self->joined_section(diag => 'item', @_);
} # end section_DIAGNOSTICS
#---------------------------------------------------------------------

=method section_CONFIGURATION_AND_ENVIRONMENT

 <module> requires no configuration files or environment variables.

=cut

sub section_CONFIGURATION_AND_ENVIRONMENT
{
  my ($self, $title) = @_;

  my ($module) = $self->required_attr($title, 'module');

  return <<"END CONFIGURATION";
\=head1 $title

$module requires no configuration files or environment variables.
END CONFIGURATION
} # end section_CONFIGURATION_AND_ENVIRONMENT
#---------------------------------------------------------------------

=method section_INCOMPATIBILITIES

 None reported.

=cut

sub section_INCOMPATIBILITIES
{
  my ($self, $title) = @_;

  "=head1 $title\n\nNone reported.\n";
} # end section_INCOMPATIBILITIES
#---------------------------------------------------------------------

=method section_BUGS_AND_LIMITATIONS

 No bugs have been reported.

=cut

sub section_BUGS_AND_LIMITATIONS
{
  my ($self, $title) = @_;

  "=head1 $title\n\nNo bugs have been reported.\n";
} # end section_BUGS_AND_LIMITATIONS
#---------------------------------------------------------------------

=attr dist

The name of the distribution that contains this module.
Required by AUTHOR.

=attr authors

An arrayref of author names (with optional email address in C<< <> >>).
Required by AUTHOR.

=attr bugtracker

An optional hashref giving the location of the distribution's public
bugtracker.  If not specified, defaults to the CPAN RT.  If present,
may have keys C<web> and C<mailto>.

=attr repository

An optional string giving the location of the distribution's public
source code repository.

=method section_AUTHOR

First, it lists the authors from the L</"authors"> attribute
(converting @ to AT in email addresses).  Then it directs bug reports
to the distribution's queue at rt.cpan.org (using the L</"dist"> attribute):

  Please report any bugs or feature requests to
  S<C<< <bug-<dist> AT rt.cpan.org> >>>,
  or through the web interface at
  L<http://rt.cpan.org/Public/Bug/Report.html?Queue=<dist>>

And, if L</"repository"> is set:

  You can follow or contribute to <dist>'s development at
  <repository>.

=cut

has qw(dist    is ro  isa Str);
has qw(authors is ro  isa ArrayRef[Str]);
has qw(bugtracker is ro  isa Maybe[HashRef]);
has qw(repository is ro  isa Maybe[Str]);

sub section_AUTHOR
{
  my ($self, $title) = @_;

  my ($dist, $authors) = $self->required_attr($title, qw(dist authors));

  my $pod = "=head1 $title\n\n";

  foreach my $authorCredit (@$authors) {
    if ($authorCredit =~ /(.*\S)\s*(<.*>)$/) {
      my ($author, $email) = ($1, $2);
      $email =~ s/@/ AT /g;
      $pod .= "$author  S<C<< $email >>>\n";
    } else {
      $pod .= "$authorCredit\n";
    }
  } # end foreach $authorCredit in @$authors

  my $bugs = $self->bugtracker || {
    mailto => "bug-$dist\@rt.cpan.org",
    web    => "http://rt.cpan.org/Public/Bug/Report.html?Queue=$dist",
  };

  if (my $mailto = $bugs->{mailto} or $bugs->{web}) {
    $pod .= "\nPlease report any bugs or feature requests\n";

    if ($mailto) {
      $mailto =~ s/@/ AT /g;
      $pod .= "to S<C<< <$mailto> >>>";
    }

    if ($bugs->{web}) {
      $pod .= "\nor " if $mailto;
      $pod .= "through the web interface at\nL<< $bugs->{web} >>";
    }
    $pod .= ".\n";
  } # end if bugtracker

  my $repo = $self->repository;
  if ($repo) {
    $repo = "L<< $repo >>" if $repo =~ /^https?:/;

    $pod .= <<"END REPOSITORY";
\nYou can follow or contribute to ${dist}'s development at
$repo.
END REPOSITORY
  } # end if $self->repository

  return $pod;
} # end section_AUTHOR
#---------------------------------------------------------------------

=attr license_notice

The complete text of the COPYRIGHT AND LICENSE section.
Required by COPYRIGHT AND LICENSE.

=method section_COPYRIGHT_AND_LICENSE

  <license_notice>

=cut

has qw(license_notice is ro  isa Str);

sub section_COPYRIGHT_AND_LICENSE
{
  my ($self, $title) = @_;

  my ($notice) = $self->required_attr($title, 'license_notice');

  #FIXME other license
  "=head1 $title\n\n$notice";
} # end section_COPYRIGHT_AND_LICENSE
#---------------------------------------------------------------------

=method section_DISCLAIMER_OF_WARRANTY

See L</"DISCLAIMER OF WARRANTY">.

=cut

sub section_DISCLAIMER_OF_WARRANTY
{
  my ($self, $title) = @_;

  return <<"END DISCLAIMER";
\=head1 $title

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENSE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.
END DISCLAIMER
} # end section_DISCLAIMER_OF_WARRANTY

#=====================================================================
# Package Return Value:

no Moose;
__PACKAGE__->meta->make_immutable;
1;

__END__

=for Pod::Loom-sort_method
section_NAME
section_VERSION
section_ATTRIBUTES
section_METHODS
section_SUBROUTINES
section_DIAGNOSTICS
section_CONFIGURATION_AND_ENVIRONMENT
section_INCOMPATIBILITIES
section_BUGS_AND_LIMITATIONS
section_AUTHOR
section_COPYRIGHT_AND_LICENSE
section_DISCLAIMER_OF_WARRANTY

=head1 DESCRIPTION

Pod::Loom::Template::Default is the default template for Pod::Loom.
It places the sections in this order:

  +  NAME
  +  VERSION
     SYNOPSIS
     DESCRIPTION
  ++ ATTRIBUTES
  ++ METHODS
  ++ SUBROUTINES
     *
  ++ DIAGNOSTICS
  +  CONFIGURATION AND ENVIRONMENT
     DEPENDENCIES
  +  INCOMPATIBILITIES
  +  BUGS AND LIMITATIONS
  +  AUTHOR
  +  COPYRIGHT AND LICENSE
  +  DISCLAIMER OF WARRANTY

Sections marked with C<+> will be provided by this template if they do
not appear in the original document.  Sections marked C<++> will be
appended to even if they do appear in the original document (if the
document contains any entries that belong in that section).

See L<Pod::Loom::Template/"Controlling the template"> for details on
how to rearrange sections and sort entries.

